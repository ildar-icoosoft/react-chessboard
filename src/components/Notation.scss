.notation {
  font-family: "Helvetica Neue";
  z-index: 3;
  position: absolute;
  font-size: 10px;
  color: rgb(240, 217, 181);
}

.fileNotation {
  @extend .notation;
  align-self: flex-end;
}
.rankNotation {
  @extend .notation;
  align-self: flex-start;
}

.lightSquare {
  color: rgb(181, 136, 99);
}
