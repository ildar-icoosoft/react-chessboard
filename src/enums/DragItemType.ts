export enum DragItemType {
  PIECE = "piece",
}
