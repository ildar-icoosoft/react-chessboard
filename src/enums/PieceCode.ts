export enum PieceCode {
  WHITE_PAWN = "wP",
  BLACK_PAWN = "bP",
  WHITE_KNIGHT = "wN",
  BLACK_KNIGHT = "bN",
  WHITE_BISHOP = "wB",
  BLACK_BISHOP = "bB",
  WHITE_ROOK = "wR",
  BLACK_ROOK = "bR",
  WHITE_QUEEN = "wQ",
  BLACK_QUEEN = "bQ",
  WHITE_KING = "wK",
  BLACK_KING = "bK",
}
