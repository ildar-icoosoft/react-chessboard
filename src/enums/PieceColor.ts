export enum PieceColor {
  WHITE = "white",
  BLACK = "black",
}
