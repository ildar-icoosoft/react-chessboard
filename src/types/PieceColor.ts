export type PieceColor = "white" | "black";
