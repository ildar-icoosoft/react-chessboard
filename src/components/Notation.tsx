import React, { FC } from "react";
import { PieceColor } from "../enums/PieceColor";
import { getFileIndex, getRankIndex, isLightSquare } from "../utils/chess";
import css from "./Notation.scss";
import classNames from "classnames";
import { DEFAULT_SQUARE_WIDTH } from "../constants/constants";

export interface NotationProps {
  coordinates: string;
  orientation: PieceColor;
  width?: number;
}

export const Notation: FC<NotationProps> = ({
  coordinates,
  orientation,
  width = DEFAULT_SQUARE_WIDTH,
}) => {
  const rankIndex: number = getRankIndex(coordinates);
  const fileIndex: number = getFileIndex(coordinates);

  const isColumn: boolean =
    orientation === PieceColor.WHITE ? rankIndex === 0 : rankIndex === 7;
  const isRow: boolean =
    orientation === PieceColor.WHITE ? fileIndex === 0 : fileIndex === 7;

  if (isColumn && isRow) {
    return renderBottomLeft(coordinates, width);
  }

  if (isColumn) {
    return renderFile(coordinates, width);
  }

  if (isRow) {
    return renderRank(coordinates, width);
  }

  return null;
};

const renderBottomLeft = (coordinates: string, width: number) => {
  return (
    <>
      {renderRank(coordinates, width)}
      {renderFile(coordinates, width)}
    </>
  );
};

const renderFile = (coordinates: string, width: number) => {
  return (
    <div
      data-testid={`notation-file-${coordinates[0]}`}
      style={{
        fontSize: `${width / 6}px`,
        paddingLeft: `${width - width / 6}px`,
      }}
      className={classNames(css.fileNotation, {
        [css.lightSquare]: isLightSquare(coordinates),
      })}
    >
      {coordinates[0]}
    </div>
  );
};

const renderRank = (coordinates: string, width: number) => {
  return (
    <div
      data-testid={`notation-rank-${coordinates[1]}`}
      style={{
        fontSize: `${width / 6}px`,
        paddingRight: `${width - width / 6}px`,
      }}
      className={classNames(css.rankNotation, {
        [css.lightSquare]: isLightSquare(coordinates),
      })}
    >
      {coordinates[1]}
    </div>
  );
};
