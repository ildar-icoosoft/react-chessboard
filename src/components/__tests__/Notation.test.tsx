import React from "react";
import TestRenderer from "react-test-renderer";
import { Notation } from "../Notation";
import { PieceColor } from "../../enums/PieceColor";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

jest.useFakeTimers();

describe("Notation", () => {
  it("Snapshot", () => {
    const testRenderer = TestRenderer.create(
      <Notation coordinates={"a1"} orientation={PieceColor.WHITE} />
    );

    expect(testRenderer.toJSON()).toMatchSnapshot();

    testRenderer.update(
      <Notation coordinates={"a3"} orientation={PieceColor.WHITE} />
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();

    testRenderer.update(
      <Notation coordinates={"d1"} orientation={PieceColor.WHITE} />
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();

    testRenderer.update(
      <Notation coordinates={"h8"} orientation={PieceColor.BLACK} />
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();

    testRenderer.update(
      <Notation coordinates={"h6"} orientation={PieceColor.BLACK} />
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();

    testRenderer.update(
      <Notation coordinates={"d8"} orientation={PieceColor.BLACK} />
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  describe("DOM structure", () => {
    it("Empty notation", () => {
      const { container } = render(
        <Notation coordinates={"b1"} orientation={PieceColor.BLACK} />
      );

      expect(container.innerHTML).toBe("");
    });

    it("Rank notation", () => {
      const { queryByTestId } = render(
        <Notation coordinates={"a5"} orientation={PieceColor.WHITE} />
      );
      expect(queryByTestId("notation-rank-5")).toBeInTheDocument();
    });

    it("File notation", () => {
      const { queryByTestId } = render(
        <Notation coordinates={"f8"} orientation={PieceColor.BLACK} />
      );
      expect(queryByTestId("notation-file-f")).toBeInTheDocument();
    });

    it("File And Rank notation", () => {
      const { queryByTestId } = render(
        <Notation coordinates={"a1"} orientation={PieceColor.WHITE} />
      );
      expect(queryByTestId("notation-rank-1")).toBeInTheDocument();
      expect(queryByTestId("notation-file-a")).toBeInTheDocument();
    });
  });
});
