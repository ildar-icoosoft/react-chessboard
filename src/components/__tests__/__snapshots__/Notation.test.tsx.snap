// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notation Snapshot 1`] = `
Array [
  <div
    className="rankNotation"
    data-testid="notation-rank-1"
    style={
      Object {
        "fontSize": "10px",
        "paddingRight": "50px",
      }
    }
  >
    1
  </div>,
  <div
    className="fileNotation"
    data-testid="notation-file-a"
    style={
      Object {
        "fontSize": "10px",
        "paddingLeft": "50px",
      }
    }
  >
    a
  </div>,
]
`;

exports[`Notation Snapshot 2`] = `
<div
  className="rankNotation"
  data-testid="notation-rank-3"
  style={
    Object {
      "fontSize": "10px",
      "paddingRight": "50px",
    }
  }
>
  3
</div>
`;

exports[`Notation Snapshot 3`] = `
<div
  className="fileNotation lightSquare"
  data-testid="notation-file-d"
  style={
    Object {
      "fontSize": "10px",
      "paddingLeft": "50px",
    }
  }
>
  d
</div>
`;

exports[`Notation Snapshot 4`] = `
Array [
  <div
    className="rankNotation"
    data-testid="notation-rank-8"
    style={
      Object {
        "fontSize": "10px",
        "paddingRight": "50px",
      }
    }
  >
    8
  </div>,
  <div
    className="fileNotation"
    data-testid="notation-file-h"
    style={
      Object {
        "fontSize": "10px",
        "paddingLeft": "50px",
      }
    }
  >
    h
  </div>,
]
`;

exports[`Notation Snapshot 5`] = `
<div
  className="rankNotation"
  data-testid="notation-rank-6"
  style={
    Object {
      "fontSize": "10px",
      "paddingRight": "50px",
    }
  }
>
  6
</div>
`;

exports[`Notation Snapshot 6`] = `
<div
  className="fileNotation"
  data-testid="notation-file-d"
  style={
    Object {
      "fontSize": "10px",
      "paddingLeft": "50px",
    }
  }
>
  d
</div>
`;
